### DISCUSSION TITLE
Dealing with additional node features and derivatives w.r.t. them

### DISCUSSION THREAD

**Discussion by shasax** [mir-group/nequip#231] [Q&A] [answered]
We are specifically dealing with atomic clusters of around 40-50 atoms (single species for now) where the atoms have something in addition to their positions as an input feature.  The output we need to learn is the cluster energy and force on each atom. Here are a couple of questions I have in this regard:

1) To provide 'position variances' as an additional input feature on each node, is it okay to use the keyword 'node_features' in 'key_mapping' in the config file? 

2) We need to train our model to obtain the forces, which are the energy derivatives w.r.t. positions, and also to obtain the derivatives w.r.t. the additional input node features we have. Is there any way/field in the config file to make this possible?

**Comment by Linux-cpp-lisp** [comment, marked as answer]
Hi @shasax ,

Sounds like an interesting project!

**Question 1:** Almost. Providing that `key_mapping` will give it as input to the model, but you also need to:
 1. Decide whether you want to concatenate your extra input to the existing one-hot input for `node_features` or replace it
 2. Clarify what the irreps for your "position variances" are, and provide them appropriately
 3. Tell the `OneHotAtomEncoding` module in the model not to override your provided `node_features`

The correct way to do this is to define a new "model builder": a function that builds/updates the model in `nequip`. By default, `nequip` uses the following list of model builders, which I'm showing here as you would specify them in the YAML config file:
```
model_builders:
  - SimpleIrrepsConfig
  - EnergyModel
  - PerSpeciesRescale
  - StressForceOutput
  - RescaleEnergyEtc
```
You will want to replace `EnergyModel` with a model builder that creates a model with the extra input feature. You do this by defining a simple Python function in a place where it can be imported (either a `setuptools` installed package or some hacks with `PYTHONPATH` like in `nequip/examples/lj`). Here's an appropriately updated version of the default model builder in https://github.com/mir-group/nequip/blob/main/nequip/model/_eng.py:
```python
def MyFancyEnergyModel(
    config, initialize: bool, dataset: Optional[AtomicDataset] = None
) -> SequentialGraphNetwork:
    logging.debug("Start building the network model")

    builder_utils.add_avg_num_neighbors(
        config=config, initialize=initialize, dataset=dataset
    )

    num_layers = config.get("num_layers", 3)

    layers = {
        # -- Encode --
        "one_hot": (OneHotAtomEncoding, dict(set_features=False)),   # !!!
        "spharm_edges": SphericalHarmonicEdgeAttrs,
        "radial_basis": RadialBasisEdgeEncoding,
        # -- Embed features --
        "chemical_embedding": AtomwiseLinear,
    }

    # add convnet layers
    # insertion preserves order
    for layer_i in range(num_layers):
        layers[f"layer{layer_i}_convnet"] = ConvNetLayer

    # .update also maintains insertion order
    layers.update(
        {
            # TODO: the next linear throws out all L > 0, don't create them in the last layer of convnet
            # -- output block --
            "conv_to_output_hidden": AtomwiseLinear,
            "output_hidden_to_scalar": (
                AtomwiseLinear,
                dict(irreps_out="1x0e", out_field=AtomicDataDict.PER_ATOM_ENERGY_KEY),
            ),
        }
    )

    layers["total_energy_sum"] = (
        AtomwiseReduce,
        dict(
            reduce="sum",
            field=AtomicDataDict.PER_ATOM_ENERGY_KEY,
            out_field=AtomicDataDict.TOTAL_ENERGY_KEY,
        ),
    )

    return SequentialGraphNetwork.from_parameters(
        shared_params=config,
        layers=layers,
        irreps_in={"node_features": "4x0e"} #!!!
    )
```
I've marked edited lines with `#!!!` comments. This is the version for if you want to _replace_ the input one-hot encoding with your input feature. There are two edits:
 1. Telling `OneHotAtomEncoding` not to override the features
 2. Specifying the expected irreps for the input features (at the very end, you'd obviously have to change the irreps to be correct)

If you want to do the version of this where you concatenate the one-hot with your features, you'd add another module to the `layers` in the builder above to do so.


**Question 2:** This is not possible from the config file alone, but is possible in a straightforward and supported way again using model builders. With the same set up as before, we can modify the existing model builder for forces (https://github.com/mir-group/nequip/blob/main/nequip/model/_grads.py) to wrap the model:
```python
def NodeFeaturesGradOutput(model: GraphModuleMixin) -> GradientOutput:
    return GradientOutput(
        func=model,
        of=AtomicDataDict.TOTAL_ENERGY_KEY,
        wrt=AtomicDataDict.NODE_FEATURES_KEY,
        out_field="node_feature_forces",
        sign=-1,  # force is the negative gradient
    )
```
Then you just add this builder to your list in your YAML config:
```
model_builders:
  - SimpleIrrepsConfig
  - mystuff.MyFancyEnergyModel
  - PerSpeciesRescale
  - ForceOutput
  - mystuff.NodeFeaturesGradOutput
  - RescaleEnergyEtc
```

**Reply by shasax** [reply]
Thanks a lot for the reply. It was indeed very helpful. Here are some questions/confusions I still have:
1) As far as I understood from the code, 'node_features' could be quantities that change with every data set, but 'node_attributes' are quantities that are expected to stay the same for all datasets (e.g. atom species encoding). Does the 'OneHotAtomEncoding' module then encode the categorical info of atom species (AtomicDataDict.ATOM_TYPE_KEY) into node_features? This is a bit counter-intuitive to my understanding above.

2) To define the new model builder 'NodeFeaturesGradOutput', I had to define first a new 'GradientOutput' module which retains the graph while doing 'torch.autograd()'. I hope this is the correct way, in the sense that it will not interfere with other parts of the codebase.

3) Can the 'out_field="node_feature_forces" from the new model builder now be used with this name in the loss function coefs in the YAML file? I basically want to train the model also wrt to these new forces.

**Reply by Linux-cpp-lisp** [reply]
Re your questions @shasax :

1. `node_features` and `node_attrs` enter into the network in different ways; the input `node_attrs` enter the network _unmodified_ at each layer while the `node_features` provided as input only directly enter the first layer.  In the default NequIP model, yes, `OneHotAtomEncoding` puts the one-hot atom type encoding both as the `node_attrs` (which are then used throughout the rest of the network) and the _initial_ `node_features` (which are replaced later by updated node features that depend on the initial node features through the convolution).
2. That should not be necessary. The `GradientOutput` module provided by NequIP (see linked files above for the imports) should do exactly that without any custom module.
3.  Yes, exactly.

**Reply by shasax** [reply]
Thanks for the reply.
Reg. point 2 above, I got an error saying that I must specify 'retain_graph=true' if I need to compute the derivatives twice. I guess it might be happening because the code goes through the network twice (once for the derivative w.r.t. node_features and then for the positions). That's why I had to define a new module. Let me know if there seems to be any better way out.

Here are some additional questions I have:
1) I notice (surprisingly) that the losses in the starting validation run (which happens before the training starts) are quite different (also in orders of magnitude) from the starting training losses that I see in the 1st batch. Are the losses computed the same way for both validation and training sets?

2) The energy and forces are scaled with the help of the 'RescaleEnergyEtc' module. However, I think I should also scale the additional node_features and the node_feature_forces. Will there a need of a custom module for that too?

3) In reality, we want to deal with systems of variable sizes (different number of atoms in different data sets). Is it okay to just provide such a collection of datasets using an npz file, or anything special might be needed for this?

**Reply by Linux-cpp-lisp** [reply]
Re `retain_graph`: according to the docs, "Defaults to the value of create_graph", which our `GradientOutput` sets to `self.training`. I suspect this means that during inference (validation), the first backwards pass is clearing the graph and then the second pass doesn't have one. One could work around this with adding some `retain_graph` stuff, but there's a better way. Instead of using `ForceOutput` and `NodeFeatureForceOutput` to add two separate `GradientOutput`, you can just add one `GradientOutput` that takes the gradient of the energy with regards to two different inputs. Do this by setting `wrt` to a list of strings and `out_field` also to a list of strings.

Then:

1. Just the loss or also the MAE, RMSE, etc.? Are are you comparing the end of epoch values or the end of validation with the first training batch?
2. That's a good point, and yes. Unfortunately at the moment I don't have this set up great: you can use the `RescaleOutput` and/or `PerSpeciesScaleShift` modules for the scaling (see the [`RescaleEnergyEtc`](https://github.com/mir-group/nequip/blob/main/nequip/model/_scaling.py) code). Basically, you want your own extra builder like RescaleEnergyEtc that calls `GlobalRescale` (like in the link) with the correct fields and a different prefix
3.  See links in https://github.com/mir-group/nequip/issues/230; use extXYZ instead

**Reply by shasax** [reply]
Thank you!
1. This issue was solved. I was writing the model builders in a wring order before and that was creating the problem.
2. Regarding this, I made an extra builder for rescaling node_feature_forces, but I am not sure if the node_features also need a rescaling. I think that the positions are also not rescaled before going on to the convolutional layers right? Would you suggest to rescale the node_features as well?
3. Thanks for this. I will have a look.

**Reply by Linux-cpp-lisp** [reply]
@shasax 

1. Great 👍  yes, all `RescaleOutput` modules must be on the outside of the model
2. Hm yes this is an interesting point. You mean the input `node_features`? Yes. In general our normalization assumes that everything is componentwise approximately zero mean and unit variance. This is enforced for geometric inputs by projecting normalized direction vectors (r "hat") onto normalized spherical harmonics, and by projecting the distances onto a normalized basis.  You definitely need to somehow make your `node_features` approximately mean zero and unit variance, but how you do that is up to you and will depend on what they are. One option is to do something like what we do with geometry, where you do normalized spherical harmonics for the angular part and then project the radial onto a normalized basis. Alternatively, if your input features are just some scalar data that you have some sense of the scale in your domain, you can just shift + scale them at the beginning of the network.
3. 👍

**Reply by shasax** [reply]
Hey again!
Thanks for all the help. The training part seems to work fine for now, at least for a fixed number of atoms in the dataset. I have now some questions for using the deployed model.

After training, we used the nequip-deploy command to get a python independent .pth file, which we want to use in a c++ script, and not directly in LAMMPS . We took some help from the 'pair_nequip' repo, which is used to integrate the use with LAMMPS. 

For a start, we are right now using a model trained using the basic NEQUIP model builders, and not including any of the new things we need, and using a dataset 'Cu_cubic.xyz' provided in the 'pair_nequip' repo. Please find attached the deployed model and the 'yaml' file for the training. Further, please also find attached the 'cpp' file we use to load and use the model.

[test_Cu.zip](https://github.com/mir-group/nequip/files/9375046/test_Cu.zip)


However, we get the following error message when the model tries to do a forward pass with the inputs:
```c++
terminate called after throwing an instance of 'c10::Error'
  what():  isTuple()INTERNAL ASSERT FAILED at "../aten/src/ATen/core/ivalue_inl.h":1916, please report a bug to PyTorch. Expected Tuple but got String
Exception raised from toTupleRef at ../aten/src/ATen/core/ivalue_inl.h:1916 (most recent call first)
```

Please find the full error message here:

![image](https://user-images.githubusercontent.com/110102212/185431273-24925ab8-92fb-4c96-9fe3-16b3597dd0d9.png)

Looking forward to your reply!

**Reply by Linux-cpp-lisp** [reply]
Hi @shasax ,

Can you figure out, either using a debugger or print statements, where in your script this failure happens?

**Reply by shasax** [reply]
Hi,
The failure happens at the point when the model tries to do a forward pass. It is not really feasible to check where the failure happens beyond that point because the control enters the libtorch libraries.

**Reply by Linux-cpp-lisp** [reply]
Hm... glancing through I don't see any obvious difference from the `pair_nequip` code. Have you successfully built and used `pair_nequip` in this environment before? Further, have you considered using `pair_nequip` directly via the [LAMMPS C API](https://docs.lammps.org/Library_execute.html)? Depending on your application, this may be much simpler by providing an existing and tested interface to NequIP along with neighborlists, atom loading utilities, neighborlists, etc.

**Reply by shasax** [reply]
Hey,
[CMakeLists.txt](https://github.com/mir-group/nequip/files/9446243/CMakeLists.txt)

It's not so easy to interface our existing codebase with the LAMMPS API actually. Is it possible if you could try compiling and running the small c++ example I sent before, with the CMakeLists.txt I attach here? 
Also, is there any small independent c++ working example that you have, for loading and using a NEQUIP generated model?

Apologies for the inconvinience, but we are really stuck on this issue it seems.

**Reply by shasax** [reply]
So, here are some updates from my side. Interfacing with a sample cpp example works fine now for a model generated without using any custom modules. However, if one tries to use the ```MyFancyEnergyModel ``` custom model builder, and then deploy the trained model, there is an error encountered on the forward pass, which is shown below: 
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  The following operation failed in the TorchScript interpreter.
Traceback of TorchScript, serialized code (most recent call last):
  File "code/__torch__/nequip/nn/_grad_output.py", line 90, in forward
        _20 = torch.append(_17, data[k])
      func1 = self.func
      data1 = (func1).forward(data, )
               ~~~~~~~~~~~~~~ <--- HERE
      of = self.of
      _21 = [torch.sum(data1[of])]
  File "code/__torch__/nequip/nn/_graph_mixin.py", line 36, in forward
    input1 = (spharm_edges).forward(input0, )
    input2 = (radial_basis).forward(input1, )
    input3 = (chemical_embedding).forward(input2, )
              ~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
    input4 = (layer0_convnet).forward(input3, )
    input5 = (layer1_convnet).forward(input4, )
  File "code/__torch__/nequip/nn/_atomwise.py", line 13, in forward
    linear = self.linear
    field = self.field
    _0 = (linear).forward(data[field], None, None, )
          ~~~~~~~~~~~~~~~ <--- HERE
    out_field = self.out_field
    torch._set_item(data, out_field, _0)
  File "code/__torch__/e3nn/o3/_linear.py", line 54, in forward
      bias0 = unchecked_cast(Tensor, bias)
    _compiled_main = self._compiled_main
    _5 = (_compiled_main).forward(features, weight0, bias0, )
          ~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
    return _5
  File "code/__torch__/torch/fx/graph_module.py", line 27, in forward
    permute = torch.permute(tensordot, [0, 2, 1])
    reshape_4 = torch.reshape(permute, [getitem_1, 32])
    return torch.reshape(reshape_4, add)
           ~~~~~~~~~~~~~ <--- HERE
class tp_forward(Module):
  __parameters__ = []

Traceback of TorchScript, original code (most recent call last):
  File "C:\Users\janhendrik\Desktop\Projects\QC\nequip\nequip\nn\_grad_output.py", line 408, in forward
            wrt_tensors.append(data[k])
        # run func
        data = self.func(data)
               ~~~~~~~~~ <--- HERE
        # Get grads
        grads = torch.autograd.grad(
  File "C:\Users\janhendrik\Desktop\Projects\QC\nequip\nequip\nn\_graph_mixin.py", line 356, in forward
    def forward(self, input: AtomicDataDict.Type) -> AtomicDataDict.Type:
        for module in self:
            input = module(input)
                    ~~~~~~ <--- HERE
        return input
  File "C:\Users\janhendrik\Desktop\Projects\QC\nequip\nequip\nn\_atomwise.py", line 56, in forward
    def forward(self, data: AtomicDataDict.Type) -> AtomicDataDict.Type:
        data[self.out_field] = self.linear(data[self.field])
                               ~~~~~~~~~~~ <--- HERE
        return data
  File "C:\Users\janhendrik\.conda\envs\nequip_QC\lib\site-packages\e3nn\o3\_linear.py", line 280, in forward
                raise RuntimeError("Biases must be provided when internal_weights = False")
            bias = self.bias
        return self._compiled_main(features, weight, bias)
               ~~~~~~~~~~~~~~~~~~~ <--- HERE
  File "<eval_with_key>.6", line 17, in forward
    permute = tensordot.permute([0, 2, 1]);  tensordot = None
    reshape_4 = permute.reshape(getitem_1, 32);  permute = getitem_1 = None
    reshape_5 = reshape_4.reshape(add);  reshape_4 = add = None
                ~~~~~~~~~~~~~~~~~ <--- HERE
    return reshape_5
RuntimeError: shape '[32]' is invalid for input of size 1760
```

So, this seems like some changes need to be made in the deployment stage as well.

The last line above says
```
RuntimeError: shape '[32]' is invalid for input of size 1760
```

So, we have 55 atoms, each with a unique node_feature. Hence, with ```num_features: 32```, it makes 55x32=1760 features. But the model somehow expects a shape of [32]. 

 Please let us know what do you think.

**Reply by Linux-cpp-lisp** [reply]
Hi @shasax ,

This looks like a much more normal error, yes :grin: 

I think `add` here which is `[32]` comes from
```
size = x.shape[:-1]
outsize = size + (irreps_out.dim,)
```
which suggests that your input features are flat instead of having shape `[55, 32]`?

**Reply by shasax** [reply]
Thank you for the help. The interfacing works fine now.
I have a question regarding the ```output metrics``` and the ```losses```. Are both of them in scaled units? The losses are computed with the scaled values I think, but are the metrics also computed for the scaled outputs, or unscaled ones?

**Reply by shasax** [reply]
Hello again,
Some weeks ago, I asked about adding a new module to get derivatives wrt the node_features. I am pasting below the combined grad and node_features_forces_rescale module I am using right now:
```
def CombinedGradOutput(model: GraphModuleMixin) -> GradientOutput:
    return GradientOutput(
        func=model,
        of=AtomicDataDict.TOTAL_ENERGY_KEY,
        wrt=[AtomicDataDict.NODE_FEATURES_KEY,AtomicDataDict.POSITIONS_KEY],
        out_field=[AtomicDataDict.NODE_FEATURE_FORCE_KEY,AtomicDataDict.FORCE_KEY],
        sign=-1,  # force is the negative gradient
    )


def NodeFeatureForcesRescale(
    model: GraphModuleMixin, config, dataset: AtomicDataset, initialize: bool
):
    return _scaling.GlobalRescale(
        model=model,
        config=config,
        dataset=dataset,
        initialize=initialize,
        module_prefix="node_feature_forces_rescale",
        default_scale=f"dataset_{AtomicDataDict.NODE_FEATURE_FORCE_KEY}_rms",
        default_shift=None,
        default_scale_keys=[AtomicDataDict.NODE_FEATURE_FORCE_KEY],
        default_shift_keys=[AtomicDataDict.NODE_FEATURE_FORCE_KEY],
        default_related_scale_keys=[],
        default_related_shift_keys=[],
    )
```

I also added the node feature forces to the AtomicDataDict, as seen above.

However, a wierd thing is noticed while predicting from the trained models. The node_feature_forces predicted, do not match the numerical differentiation wrt node_features of the predicted energy landscape (which should not be the case as the model also tries to find the derivative using the same network which outputs the energy). The issue has something to do with one of the modules above. Could you have a quick glance at them, and suggest something else which might be going wrong?
    
  Thanks in advance!

**Reply by Linux-cpp-lisp** [reply]
>  I have a question regarding the output metrics and the losses. Are both of them in scaled units? The losses are computed with the scaled values I think, but are the metrics also computed for the scaled outputs, or unscaled ones?

Output metrics are always in the original units of the training data.

**Reply by Linux-cpp-lisp** [reply]
Can you elaborate on what numerical differentiation you are comparing to?

Something might be wrong with the scaling of the node feature forces... I may have given you the wrong advice there.  Since they are a derivative of the energy, by linearity, they also have to share that same scaling--- otherwise it doesn't make any sense.  So really, `NODE_FEATURE_FORCES` should be added to `ALL_ENERGY_KEYS` so that it gets rescaled in tandem with the energy, forces, stress, etc. by the normal force RMS. 

That still doesn't quite resolve the issue of the scale of the _input_ node features, but that's something that has to be solved in preprocessing of the dataset or by scaling them to normal by some constant as a first step in the model inference pipeline. (Which is the opposite of scaling the output node feature forces, as is happening here.)

**Reply by shasax** [reply]
So, by numerical derivative, I mean that let's say we predict the model outputs on an input of same positions but uniformly spaced node_features. 
In that case, at any input point, the numerical derivative of the predicted energy landscape wrt node_features (delta E / delta node_features) should be similar to the node_feature_forces we observe in the predictions. But these two are very much apart, which points to something fishy. 

Ah, if that is the case for node_feature_forces scaling, I will try that. Previously, I was trying to scale them separately with their RMS value. 

Regarding scale of input node features, I do them in the data preprocessing step now. 

Please let me know any other implementations you might suggest to be wrong currently. Thanks again for the help :)

**Reply by Linux-cpp-lisp** [reply]
Got it, sounds correct--- I think it is the difference in the output scaling being inconsistent with the derivative then.  Since d (scaling * E) / d node_features = scaling * dE/d node_features, but not equal to scaling2 * dE/d node_features.

> Regarding scale of input node features, I do them in the data preprocessing step now.

Great, then you should be fine. (Be careful of course to scale the forces back out in post-processing when you do inference then.)

> Please let me know any other implementations you might suggest to be wrong currently. Thanks again for the help :)

All sounds good to me right now 👍

**Reply by shasax** [reply]
Hello,
Thank you for all the support.
I am reopening this thread after a long time. So just to recap a bit, I have been working with NEQUIP after tweaking the ```EnergyModel``` slightly, where I replaced the one-hot atom encoding with a list of ```double```-type node features. I previously received help from your side, which made me define my own energy model as shown below:
```
def MyFancyEnergyModel(
    config, initialize: bool, dataset: Optional[AtomicDataset] = None
) -> SequentialGraphNetwork:
    logging.debug("Start building the network model")

    builder_utils.add_avg_num_neighbors(
        config=config, initialize=initialize, dataset=dataset
    )

    num_layers = config.get("num_layers", 3)

    layers = {
        # -- Encode --
        "one_hot": (OneHotAtomEncoding, dict(set_features=False)),   # !!!
        "spharm_edges": SphericalHarmonicEdgeAttrs,
        "radial_basis": RadialBasisEdgeEncoding,
        # -- Embed features --
        "chemical_embedding": AtomwiseLinear,
    }

    # add convnet layers
    # insertion preserves order
    for layer_i in range(num_layers):
        layers[f"layer{layer_i}_convnet"] = ConvNetLayer

    # .update also maintains insertion order
    layers.update(
        {
            # TODO: the next linear throws out all L > 0, don't create them in the last layer of convnet
            # -- output block --
            "conv_to_output_hidden": AtomwiseLinear,
            "output_hidden_to_scalar": (
                AtomwiseLinear,
                dict(irreps_out="1x0e", out_field=AtomicDataDict.PER_ATOM_ENERGY_KEY),
            ),
        }
    )

    layers["total_energy_sum"] = (
        AtomwiseReduce,
        dict(
            reduce="sum",
            field=AtomicDataDict.PER_ATOM_ENERGY_KEY,
            out_field=AtomicDataDict.TOTAL_ENERGY_KEY,
        ),
    )

    return SequentialGraphNetwork.from_parameters(
        shared_params=config,
        layers=layers,
        irreps_in={"node_features": "4x0e"} #!!!
    )
```

However, now I would want to concatenate the one-hot atom encoding with my double-style node features. Can it be done using the same model builder? Or does this require significant changes? 

What I believe would work is that I concatenated the one-hot tensor to my input node features inside the ```forward``` definition of class ```OneHotAtomEncoding```, by using:
```
total_node_features = torch.cat((data[AtomicDataDict.NODE_FEATURES_KEY],one_hot),1)
data[AtomicDataDict.NODE_FEATURES_KEY] = total_node_features
```
Is something else also needed?

Also, I have a more extended version of the ```ForceOutput``` model builder, which computes the gradient w.r.t. the positions and the node features. This is shown below.

```
def CombinedGradOutput(model: GraphModuleMixin) -> GradientOutput:
    if AtomicDataDict.FORCE_KEY in model.irreps_out:
        raise ValueError("This model already has force outputs.")
    return GradientOutput(
        func=model,
        of=AtomicDataDict.TOTAL_ENERGY_KEY,
        wrt=[AtomicDataDict.NODE_FEATURES_KEY,AtomicDataDict.POSITIONS_KEY],
        out_field=[AtomicDataDict.NODE_FEATURE_FORCE_KEY,AtomicDataDict.FORCE_KEY],
        sign=-1,  # force is the negative gradient
    )
```
 
 As the gradient-output builder is usually on the outside of the one-hot encoding, the node features are not yet concatenated with the one-hot features when the gradient is computed. But this also is something desired as I do not want derivatives w.r.t. the one-hot features. 

Does this sound alright, or something needs to be changed also in the ```CombinedGradOutput```?

Some help would be highly appreciated!

**Reply by Linux-cpp-lisp** [reply]
Hi @shasax ,

I think you're right that a modification / replacement of `OneHotAtomEncoding` to allow concatenating to existing features.  Note, however, that you will also need to report the correct concatenation of the irreps of the one hot features and the incoming other node features in `irreps_out` of the `OneHotAtomEncoding`. (If I remember right, `o3.Irreps`, being `Tuple`s, can be concatenated with `+`.)  (Also can pay to be careful with dimensions, usually better to use `-1` as the concat dim rather than `1` just to be on the safe side and consistent with other code in the framework.) 

Regarding gradients, the `GradientOutput` happens around the entire `SequentialGraphNetwork`, so PyTorch can understand that the concatenated features are derived from the original node features and should propagate gradients accordingly.  (In other words, you are asking for the gradient w.r.t. the node features _when they enter the beginning of the GradientOutput_, not at any later redefinition.)
